src/main/java/pl/sdacademy/carrental/services/ClientService.java

```diff
@@ -0,0 +1,88 @@
+import pl.sdacademy.carrental.repositories.ClientRepository;
+
+import java.util.List;
+
+@Service
+@Transactional
+public class ClientService {
+
+    private final ClientRepository clientRepository;
+    private final AddressRepository addressRepository;
+
+    public ClientService(ClientRepository clientRepository, AddressRepository addressRepository) {
+        this.clientRepository = clientRepository;
+        this.addressRepository = addressRepository;
+    }
+
+    public List<Client> getAll() {
+        return clientRepository.findAll();
+    }
+
+    public void createClient(final ClientForm clientForm) {
+
+        final Address address = createAddressFromClientForm(clientForm);
+
+        final Client client = Client.builder()
+                .firstName(clientForm.getFirstName())
+                .lastName(clientForm.getLastName())
+                .email(clientForm.getEmail())
+                .address(address)
+                .phoneNumber(clientForm.getPhoneNumber())
+                .build();
+        clientRepository.save(client);
+    }
+
+    public void delete(final Long id) {
+        final Client client = clientRepository.findById(id).orElseThrow();
+        clientRepository.delete(client);
+    }
+
+    public void updateClient(final Long id, final ClientForm clientForm) {
+
+        final Address address = createAddressFromClientForm(clientForm);
+
+        final Client client = clientRepository.findById(id).orElseThrow();
+        client.setFirstName(clientForm.getFirstName());
+        client.setLastName(clientForm.getLastName());
+        client.setAddress(address);
+        clientRepository.save(client);
+    }
+
+    private Address createAddressFromClientForm(final ClientForm clientForm){
+        final Address address = Address.builder()
+                .city(clientForm.getCity())
+                .street(clientForm.getStreet())
+                .building(clientForm.getBuilding())
+                .apartment(clientForm.getApartment())
+                .zipCode(clientForm.getZipCode())
+                .build();
+        return addressRepository.save(address);
+    }
+    // I don't know if saving both branch and client address to the same repository is a good idea
+    // Split, refactor AddressRepository to ClientAddressRepository and BranchAddressRepository?
+    // same with Address entity? What do you think?
+
+    public ClientForm getById(final Long id) {
+        final Client client = clientRepository.findById(id).orElseThrow();
+        final Address address = client.getAddress();
+        return ClientForm.builder()
+                .firstName(client.getFirstName())
+                .lastName(client.getLastName())
+                .city(address.getCity())
+                .street(address.getStreet())
+                .apartment(address.getApartment())
+                .zipCode(address.getZipCode())
+                .phoneNumber(client.getPhoneNumber())
```

Comment:
brakuje `.building(address.getBuilding())` i `.email(client.getEmail())`

Reply:
Dołożyłem IOException w miejscu gdzie trzeba należało dołożyć konkretny, ale nie jestem 100% pewny, czy jest to najodpowiedniejszy exception jaki można było wstawić.

Reply:
ten IOE tutaj średnio, lepiej dać dedykowany, a jeśli nie chcesz teraz tego robić, to zmienić metodę z `.findById(id)` na `.getOne(id)` - ona rzuca `EntityNotFoundException`